### PR TITLE
Project local plugins

### PR DESCRIPTION
**Pull request by asutoshpalai**
If the user is in an app directory the plugin is installed in `.bundle/` inside the app root.

If the user is in not in any app directory the plugin is installed in the `.bundle/` in the user's home.

### CODE REVIEW

File: lib/bundler/plugin.rb [hunk 1]
```diff
@@ -128,6 +144,14 @@ def installed?(plugin)
       Index.new.installed?(plugin)
     end
 
+    # Used by specs
+    def reset!
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: `Bundler.reset!` should call this

**Comment by segiddins** [comment]
This seems like it might be a bit confusing -- silently switching behavior based upon the current working director

**Comment by asutoshpalai** [comment]
Any other suggestions for project level plugins.

An alternative I thought is to make `source` plugins project level and `command` plugins global.

**Comment by segiddins** [comment]
@asutoshpalai maybe the way `bundle config` does it? defaults to local, has `--local` and `--global`?

**Comment by asutoshpalai** [comment]
Yes, right!
But what about the cases when the user is not in any app directory and tries to install/use plugin (command plugins)? Should we fail or switch silently (or verbosly)?

**Comment by segiddins** [comment]
@indirect thoughts?

**Comment by indirect** [comment]
Source plugins should definitely be project level—they only apply to a single gemfile at a time.

My (admittedly vague) general idea was that command plugins would be global, and installed via `gem install` rather than listed in a gemfile—but I don't know how that interacts with command plugins that are used by a specific project. It seems like `bundle grep` should be a global plugin, while `bundle audit` should be a project-level plugin. Maybe we should support both, for different kinds of commands?

**Comment by asutoshpalai** [comment]
How about this:
Everything thing is in general project level. When in a project **only commands** get _merged_ from global level (can't see how it harms since they don't modify bundler's flow).

When not in any project, the global commands are available.

**Comment by indirect** [comment]
@segiddins r?

File: lib/bundler.rb [hunk 2]
```diff
@@ -393,6 +393,7 @@ def reset!
       rubygems.undo_replacements
       rubygems.reset
       @rubygems = nil
+      Plugin.reset!
```
Comment (segiddins) [inline lib/bundler.rb]: this needs to be executed regardless of whether theres a `rubygems` or not

File: spec/bundler/plugin/index_spec.rb [hunk 3]
```diff
@@ -75,6 +79,20 @@
     end
   end
 
+  describe "global index" do
+    before do
+      Dir.chdir tmp
```
Comment (segiddins) [inline spec/bundler/plugin/index_spec.rb]: this should use the block form of `chdir`

File: spec/support/path.rb [hunk 4]
```diff
@@ -81,12 +81,12 @@ def bundler_path
       Pathname.new(File.expand_path("../../../lib", __FILE__))
     end
 
-    def plugin_root(*args)
-      home ".bundle", "plugin", *args
+    def global_plugin_gems(*args)
```
Comment (segiddins) [inline spec/support/path.rb]: `gem` in the singular

File: spec/support/path.rb [hunk 5]
```diff
@@ -81,12 +81,12 @@ def bundler_path
...
+      home ".bundle", "plugin", "gems", *args
     end
 
-    def plugin_gems(*args)
-      plugin_root "gems", *args
+    def local_plugin_gems(*args)
```
Comment (segiddins) [inline spec/support/path.rb]: singular `gem`

**Comment by indirect** [comment]
@segiddins how's it looking now, r?

**Comment by segiddins** [comment]
@homu r+

**Pull request merged**